synapse/config/_base.py

```diff
@@ -727,6 +766,20 @@ def generate_missing_files(
     ) -> None:
         self.invoke_all("generate_files", config_dict, config_dir_path)
 
+    def reload_config_section(self, section_name: str) -> None:
+        """Reconstruct the given config section, leaving all others unchanged.
+
+        :raises ValueError: if the given `section` does not exist.
+        :raises ConfigError: for any other problems reloading config.
+        """
+        existing_config: Optional[Config] = getattr(self, section_name, None)
+        if existing_config is None:
+            raise ValueError(f"Unknown config section '{section_name}'")
+        logger.info("Reloading config section '%s'", section_name)
+
+        new_config_data = read_config_files(self.config_files)
+        existing_config.read_config(new_config_data)
```

Comment:
What happens if we error half way through reading the config? I.e. if someone has put an invalid value in there? Do we end up in a messed up state where we've reloaded *some* of the config?

I guess ideally we should be reading the config into a *new* config class and then replacing the existing config with that once we've successfully read it?

Reply:
> Do we end up in a messed up state where we've reloaded _some_ of the config?

We will, yes. I don't think that's the end of the world, because you can correct and re-reload the config. But it would be nice for this to be an atomic operation. I'll do this shortly

Reply:
Note however that the CacheConfig in particular is naughty. When it reads in config it adjusts global state in its enclosing module. And it automatically re-evaluates cache sizes. I'll have to change this too.

Reply:
Oh that's annoying :(

Reply:
> Could use somebody to sanity check that it works though.

The reloading seems to work fine. Not sure how to check that the caches themselves have the appropriate size from the diffs?